src/linter/block.go

```diff
@@ -1514,33 +1518,31 @@ func (b *BlockWalker) handleAssignList(list *ir.ListExpr, rhs ir.Node) {
 	// and get stack allocation which will help to avoid the unwanted heap allocs.
 	// Hint: only const (literal) size hints work for this.
 	// Hint: check the compiler output to see whether elemTypes "escape" or not.
-	//
-	// We store meta.Type instead of string to avoid the need to do strings.Join
-	// when we want to create a TypesMap.
-	var elemTypes []meta.Type
-	var shapeType string
-	typ.Iterate(func(typ string) {
+	var elemTypes = make(meta.RawTypesMap, typ.Len())
+	var shapeType meta.Type
+
+	typ.Iterate(func(typ meta.Type) {
 		switch {
-		case meta.IsShapeType(typ):
+		case typ.IsShape():
```

Comment:
So is this basically the problem that you are trying to solve by refactoring everything to concrete types? Why not put that into the PR description?

Reply:
Initially, I wanted to avoid low-level interactions, such as here:

```go
	left.Iterate(func(typ string) {
		if typ[0] == meta.WArrayOf {
			hasAtLeastOneArray = true
		}
		if typ[0] == '\\' {
			hasAtLeastOneClass = true
		}
		newMap.AppendString(typ)
	})

	right.Iterate(func(typ string) {
		if typ[0] == meta.WArrayOf && meta.UnwrapArrayOf(typ) == "mixed" && hasAtLeastOneArray {
			return
		}
		if typ == "object" && hasAtLeastOneClass {
			return
		}
		newMap.AppendString(typ)
	})
```

So that we can conveniently work with individual types without going into implementation details and see all the methods that are available for them, and not a certain set of functions, such as `IsArrayType`, which for some reason is in the` meta` package.